test/spectests/spectests.cpp

```diff
@@ -0,0 +1,294 @@
+                    continue;
+                }
+
+                pass();
+            }
+            else if (type == "assert_invalid")
+            {
+                const auto filename = cmd.at("filename").get<std::string>();
+                const auto wasm_binary = load_wasm_file(path, filename);
+                try
+                {
+                    fizzy::parse(wasm_binary);
+                }
+                catch (fizzy::parser_error const&)
+                {
+                    pass();
+                    continue;
+                }
+
+                fail("Invalid module parsed successfully. Expected error: " +
+                     cmd.at("text").get<std::string>());
+            }
+            else
+                skip("Unsupported command type");
+        }
+
+        log(std::to_string(passed + failed + skipped) + " tests ran from " +
+            path.filename().string() + ".\n  PASSED " + std::to_string(passed) + ", FAILED " +
+            std::to_string(failed) + ", SKIPPED " + std::to_string(skipped) + ".\n");
+    }
+
+private:
+    std::optional<fizzy::execution_result> invoke(const json& action)
+    {
+        const auto func_name = action.at("field").get<std::string>();
+        const auto func_idx = fizzy::find_exported_function(instance.module, func_name);
+        if (!func_idx.has_value())
+        {
+            skip("Function '" + func_name + "' not found.");
+            return std::nullopt;
+        }
+
+        std::vector<uint64_t> args;
+        for (const auto& arg : action.at("args"))
+        {
+            const auto arg_type = arg.at("type").get<std::string>();
+            uint64_t arg_value;
+            if (arg_type == "i32")
+                arg_value = json_to_value<int32_t>(arg.at("value"));
+            else if (arg_type == "i64")
+                arg_value = json_to_value<int64_t>(arg.at("value"));
+            else
+            {
+                skip("Unsupported argument type '" + arg_type + "'.");
+                return std::nullopt;
+            }
+            args.push_back(arg_value);
+        }
+
+        return fizzy::execute(instance, *func_idx, std::move(args));
+    }
+
+    void pass()
+    {
+        ++passed;
+        std::cout << "PASSED\n";
+    }
+
+    void fail(std::string_view message)
+    {
+        ++failed;
+        std::cout << "FAILED " << message << "\n";
+    }
+
+    void skip(std::string_view message)
+    {
+        ++skipped;
+        std::cout << "SKIPPED " << message << "\n";
+    }
+
+    void log(std::string_view message) const { std::cout << message << "\n"; }
+
+    void log_no_newline(std::string_view message) const { std::cout << message << std::flush; }
+
+    fizzy::Instance instance;
+    int passed = 0;
+    int failed = 0;
+    int skipped = 0;
+};
+
+void run_tests_from_dir(const fs::path& path)
+{
+    std::vector<fs::path> subdirs;
+    std::vector<fs::path> files;
+
+    for (const auto& e : fs::directory_iterator{path})
```

Comment:
Have you considered `recursive_directory_iterator`?

Reply:
Maybe `copy_if(recursive_directory_iterator{}, vector, is_json_file)`, then `sort()`.

Reply:
With `copy_if` it's `begin(iterator)` + `end(iterator)` + `back_inserter` + lambda, so I preferred three lines of good old range for.